File: Review-14-Python-Builder-Pattern/Example-3/house.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable, List, Optional, Self

from room import Room


DEFAULT_NAME = "House"


class House:
    def __init__(self, *, name: str, rooms: list[Room]):
        """
        This is the Default constructor. Start with the name set to "House" and
        an empty set of rooms (i.e., zero rooms).
        """

        self.__name: str = name
        self.__rooms: list[Room] = rooms

    @property
    def name(self):
        """
        Get the name using a property.

        @property must come before @...setter...
        """

        return self.__name

    def __len__(self):
        """
        Return the number of rooms in this House.
        """

        return len(self.__rooms)

    def flooring_cost_metrics(self) -> tuple[float, float]:
        #  total = sum(map(lambda room: room.flooring_cost(), self))
        total = sum(room.flooring_cost() for room in self)
        avg = total / len(self)

        return total, avg

    def __iter__(self):
        """
        Wrapper around `list.__iter__()`.
        """

        return iter(self.__rooms)

    def __str__(self) -> str:
        """
        This is the equivalent of overloading:
          - `operator<<` in C++
          - `toString` in Java
          - `Display::fmt` in Rust
        """

        total, avg = self.flooring_cost_metrics()

        # TODO: replace '+" with '"\n".join(...)'
        return (
            f"--------{self.name:}--------\n"
            + "\n".join(str(room) for room in self)
            + "------------------------------\n"
            + f"Total Cost   : $ {total:.2f}\n"
            + f"Avg Room Cost: $ {avg:.2f}\n"
        )

    def __eq__(self, rhs) -> bool:
        """
        This is the equivalent of overloading:
          - `operator==` in C++
          - `equals` in Java
          - `eq` in Rust
        """

        # TODO: Add instanceof check

        if self.name != rhs.name:
            return False

        if len(self) != len(rhs):
            return False

        # Python handles the traversal for us!
        return self.__rooms == rhs.__rooms


@dataclass
class HouseBuilder:
    name: Optional[str] = None
    the_rooms: list[Room] = field(default_factory=list)

    def with_name(self, nme: str) -> Self:
        """
        Set house name. This is optional.
        """
        self.name = nme

        return self

    def with_room(self, room: Room) -> Self:
        """
        Add a single room
        """
        self.the_rooms.append(room)

        return self

    def with_rooms(self, new_rooms: Iterable[Room]) -> Self:
        """
        Add multiple rooms from any iterable object
        """

        self.the_rooms.extend(new_rooms)

        return self

    def build(self) -> House:
        if not self.the_rooms:
            raise ValueError("A House must have at least one room.")

        # If a name was supplied use it... otherwise supply the default.
        if self.name:
            return House(name=self.name, rooms=self.the_rooms)

        return House(name=DEFAULT_NAME, rooms=self.the_rooms)

File: 14-Rust-Builder-Pattern/Example-6/src/house.rs
use std::fmt::{self, Display, Formatter};
use std::vec::Vec;

use crate::error::{HouseError, HouseErrorWithState};
use crate::room::Room;

type Collection = Vec<Room>;

#[derive(Clone, Debug, Default)]
pub struct House {
    name: String,
    rooms: Collection,
}

impl House {
    pub fn builder() -> HouseBuilder<NoRooms> {
        HouseBuilder::new()
    }

    /// Get the name using a traditional accessor.
    ///
    pub fn get_name(&self) -> &str {
        &self.name
    }

    /// Return the number of rooms in this House.
    ///
    pub fn len(&self) -> usize {
        self.rooms.len()
    }

    /// Determine whether this House is empty (i.e. `self.len() == 0).
    ///
    pub fn is_empty(&self) -> bool {
        self.rooms.is_empty()
    }

    /// Wrapper around `Collection<Room>.iter()`.
    ///
    pub fn iter(&self) -> impl Iterator<Item = &Room> {
        self.rooms.iter()
    }

    /// Wrapper around `Collection<Room>.iter_mut()`.
    ///
    pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut Room> {
        self.rooms.iter_mut()
    }

    pub fn flooring_metrics(&self) -> (f64, f64) {
        let total = self.rooms.iter().map(|room| room.flooring_cost()).sum();

        let avg = total / (self.len() as f64);

        (total, avg)
    }
}

impl Display for House {
    /// This is the equivalent of overloading:
    ///   - `operator<<` in C++
    ///   - `toString` in Java
    ///   - `__str__` in Python
    ///
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        writeln!(f, "--------{}--------\n", self.name)?;

        for room in self.rooms.iter() {
            writeln!(f, "{}", room)?;
        }

        let (total, avg) = self.flooring_metrics();

        writeln!(f, "------------------------------")?;
        writeln!(f, "Total Cost   : $ {:.2}", total)?;
        writeln!(f, "Avg Room Cost: $ {:.2}", avg)
    }
}

impl PartialEq for House {
    /// This is the equivalent of overloading:
    ///   - `operator==` in C++
    ///   - `equals` in Java
    ///   - `__eq__` in Python
    fn eq(&self, rhs: &Self) -> bool {
        if self.name != rhs.name {
            return false;
        }

        if self.len() != rhs.len() {
            return false;
        }

        // We are guaranteed to have the same number of rooms
        for (lhs_room, rhs_room) in self.iter().zip(rhs.iter()) {
            if lhs_room != rhs_room {
                return false;
            }
        }

        return true;
    }
}

//------------------------------------------------------------------------------
#[derive(Default, Debug, PartialEq)]
pub struct NoRooms;

#[derive(Debug, PartialEq)]
pub struct HouseBuilder<SR> {
    name: String,
    rooms: SR,
}

/// The HouseBuilder's state is defined by one generics
///   - SR - (State Rooms) whether at least one room has been added
///
impl HouseBuilder<NoRooms> {
    pub fn new() -> Self {
        HouseBuilder {
            name: "House".to_owned(),
            rooms: Default::default(),
        }
    }
}

impl<SR> HouseBuilder<SR> {
    pub fn with_name(mut self, nme: &str) -> Self {
        self.name = nme.to_owned();

        self
    }
}

impl HouseBuilder<NoRooms> {
    pub fn with_room(self, first_room: Room) -> HouseBuilder<Vec<Room>> {
        HouseBuilder {
            name: self.name,
            rooms: vec![first_room],
        }
    }

    pub fn with_rooms(
        self,
        first_rooms: Vec<Room>,
    ) -> Result<HouseBuilder<Vec<Room>>, HouseErrorWithState<Self>> {
        match first_rooms.len() {
            0 => Err(HouseErrorWithState {
                the_error: HouseError::ZeroRooms,
                the_builder: self,
            }),
            _ => Ok(HouseBuilder {
                name: self.name,
                rooms: first_rooms,
            }),
        }
    }
}

impl HouseBuilder<Vec<Room>> {
    pub fn with_room(mut self, another_room: Room) -> Self {
        self.rooms.push(another_room);

        self
    }

    pub fn with_rooms(mut self, mut more_rooms: Vec<Room>) -> Self {
        self.rooms.append(&mut more_rooms);

        self
    }

    pub fn build(self) -> House {
        House {
            name: self.name.to_owned(),
            rooms: self.rooms,
        }
    }
}
